### PR TITLE
Auto download wallet statement after finish genereating

### PR DESCRIPTION
**Pull request by bernhardoj**
### Details

Currently, user needs to press twice the download button to download a wallet statement. We want to auto-download it after finish generating.

### Fixed Issues

$ https://github.com/Expensify/App/issues/50828
PROPOSAL: https://github.com/Expensify/App/issues/50828#issuecomment-2415762459

### Tests

Same as QA Steps
- [x] Verify that no errors appear in the JS console

### Offline tests

Same as QA Steps
### QA Steps

1. Open a wallet statement page
2. Press on the download button
3. Verify a loader is shown and the statement is downloaded afterward
- [x] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

https://github.com/user-attachments/assets/36a97ab2-d1e4-414c-87ad-57599a5b7747

</details>

<details>
<summary>Android: mWeb Chrome</summary>

https://github.com/user-attachments/assets/7db1fc2b-03fb-486b-86da-561ccb705b84

</details>

<details>
<summary>iOS: Native</summary>

https://github.com/user-attachments/assets/eafbb6bf-51c6-48ff-a97f-c41076f5e23b

</details>

<details>
<summary>iOS: mWeb Safari</summary>

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/user-attachments/assets/51bb0356-a1a5-4a30-a73b-f7bef478cd68

</details>

<details>
<summary>MacOS: Desktop</summary>

https://github.com/user-attachments/assets/406d697d-c14d-44fe-b72b-03e8a73ae229

</details>

### CODE REVIEW

**Comment by bernhardoj** [comment]
Looks like the download doesn't work in iOS mWeb (on `main` too).

https://github.com/user-attachments/assets/886f849f-d6ae-4f92-9f35-51b9e6592360

**Review by fedirjh — APPROVED** [review]
Looks good to me.

cc @Expensify/design

**Comment by dannymcclain** [comment]
Looking good. Is there any way we can keep the spinner going until the download thing actually happens/pops up? This is definitely much better than it was, but it seems like there's a couple seconds after the spinner goes away and before the download thingy pops up that makes me think "Ok wait, did it do it or did it mess up?"

**Comment by shawnborton** [comment]
Yeah I think that's a fair point Danny.

**Comment by dubielzyk-expensify** [comment]
Agree with that. Would make it clearer

**Comment by bernhardoj** [comment]
Done.

https://github.com/user-attachments/assets/9c8b0961-281e-41ca-bcf8-953b1c9a1ec2

**Comment by shawnborton** [comment]
I think that looks better - it does seem like the continuity of the loading circle breaks slightly (as if it resets quickly) but I think that's something we can probably not worry about.

**Comment by dannymcclain** [comment]
>it does seem like the continuity of the loading circle breaks slightly (as if it resets quickly) but I think that's something we can probably not worry about.

Yeah this isn't _super_ noticeable (I mean... You and me noticed it but 😅) and I think we can not worry about it.

It feels much better than before.

**Comment by shawnborton** [comment]
Agree! Lets Get This Merged.

**Review by fedirjh — COMMENTED** [review]
> it does seem like the continuity of the loading circle breaks slightly

@bernhardoj I think we can address that with the suggested changes.

**Comment by bernhardoj** [comment]
Done. Please check again.

File: src/pages/wallet/WalletStatementPage.tsx [hunk 1]
```diff
@@ -49,23 +45,35 @@ function WalletStatementPage({walletStatement, route}: WalletStatementPageProps)
...
+
+    // eslint-disable-next-line rulesdir/prefer-early-return
+    useEffect(() => {
+        // If the statement generate is complete, download it automatically.
+        if (prevIsWalletStatementGenerating && !isWalletStatementGenerating) {
+            if (walletStatement?.[yearMonth]) {
+                processDownload();
+            } else {
+                setIsDownloading(false);
+            }
```
[the remark this answers is not in the document]
Comment (bernhardoj) [reply src/pages/wallet/WalletStatementPage.tsx:74]: Just in case the generate fails.

**Review by fedirjh — APPROVED** [review]
Looks good :shipit: 

https://github.com/user-attachments/assets/c0045a02-2dd4-48bc-9709-9d5c511476eb

https://github.com/user-attachments/assets/dd8c2e9b-bdb9-46ad-9a0e-6a05678709b9

**Review by carlosmiceli — APPROVED** [review]

**Comment by mvtglobally** [comment]
We are not able to see statements in Wallet. Is there a way to trigger it in QA or can you validate internally ?

**Comment by bernhardoj** [comment]
@mvtglobally sorry, I should've put it on the QA Steps, but you can try accessing the wallet statement page directly.
`https://staging.new.expensify.com/statements/{yyyy}{mm}`
`https://staging.new.expensify.com/statements/202404`

**Comment by Julesssss** [comment]
Hey @carlosmiceli @bernhardoj, I can't figure out how to access the statement page using the direct link. Please can you test and verify this on staging?

Regardless in the meantime I will skip this from the checklist because it is a hidden feature that shouldn't block deploy.

**Comment by bernhardoj** [comment]
Yes, I see that too, but you can still download it. Here is the recording on staging.

https://github.com/user-attachments/assets/1be2da1e-5cbd-4dbf-91ce-698c3a2e2f2b

**Comment by Julesssss** [comment]
Ah thanks. After realising this wasn't live for users I was less concerned and removed the blocker, but thanks for pointing this out 👍

**Pull request merged**